docs/proposals/edge-node-upgrade.md

```diff
@@ -0,0 +1,273 @@
+  creation-date: 2022-06-26
+  last-updated: 2022-06-26
+  status: implementable
+---
+
+# Edge Node Upgrade
+
+## Motivation
+
+Edge node upgrade management is a key feature required for upgrading edge nodes from remote cloud side in edge computing.
+This proposal addresses how can we upgrade edge nodes from the cloud, and synchronize the edge node upgrade result status between edge nodes and cloud.
+
+### Goals
+
+Edge node upgrade management must:
+* provide APIs for upgrading edge nodes from the cloud.
+* synchronize the edge node upgrade result between cloud and edge nodes.
+
+## Proposal
+We propose using Kubernetes [Custom Resource Definitions (CRDs)](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)
+to describe Upgrade metadata/status and a controller to synchronize Upgrade between edge and cloud.
+
+### Use Cases
+
+* Describe Upgrade properties.
+  * Users can describe Upgrade properties and access mechanisms to interact with / control the Upgrade.
+* Perform CRUD operations on Upgrade from cloud.
+  * Users can create, update and delete Upgrade metadata from the cloud via the CRD APIs exposed by the Kubernetes API server.
+* Report Upgrade properties values.
+  * Edge nodes can report upgrade result status.
+
+
+## Upgrade Controller Design
+The Upgrade controller starts two separate goroutines called  `upstream` controller and `downstream` controller.
+These are not separate controllers as such but named here for clarity.
+The job of the downstream controller is to synchronize the Upgrade updates from the cloud to the edge node.
+The job of the upstream controller is the reverse.
+
+
+## Synchronizing Upgrade
+
+The below illustrations describe the flow of events that would occur when Upgrade property values are updated from the cloud/edge.
+
+- Users create/modify the Upgrade CR directly using `kubectl`.
+- Upgrade Controller Downstream receive events from K8s APIServer, and then store it in local cache
+- Upgrade Controller Downstream checks whether the upgrade is completed(ReUpgrade is false and all or partial node upgrade is completed),
+  if so, we don't need to send upgrade message to edge node
+- Upgrade Controller Downstream call K8s api to get node list using NodeNames or LabelSelector specified in CR, and
+  filter out nodes that don't meet upgrade requirements(1. edge node already on the desired version. 2. not edge node,
+  without label {"node-role.kubernetes.io/edge": ""} 3. remove duplicated nodes)
+- Upgrade Controller Downstream send Upgrade message to each edge node. And new a Goroutine to process timeout
+  if we don't receive edge node upgrade response, if not, downstream will construct a timeout response message and send it to
+  Upstream to let Upstream update Upgrade status. And patch update Upgrade status state to Upgrading state.
+  And mark edge node unschedulable. If the ReUpgrade flag is true, we will patch update it from true to false, to avoid always reupgrading.
```

Comment:
waht does ReUpgrade mean?  It means that when the edge node upgrade failed, it can be retried?

Reply:
yes, that's one case. And if we want to use the existing Upgrade CR to upgrade node to another version, we can set ReUpgrade to true and modify the version, so that will trigger another upgrade operation